feat(minor): Implement __getitem__ in Base Document

gavindsouza opened the pull request:
This addition allows for syntactic sugars like writing `doc["name"]` to access document name. This addition is solely for the sake of code simplicity. Since dictionaries in Python follow this standard, we can write code that works for both Vanilla Dicts, Frappe Dicts & all Document objects.

This allows for accessing the document attributes in one more way now.

<img width="453" alt="Screenshot 2021-11-01 at 12 47 18 PM" src="https://user-images.githubusercontent.com/36654812/139636245-24ab3e42-b01e-4af9-89fd-9c89fc3374a8.png">


### Motivation

Plainly, this change is driven by the following error that comes up when you try downloading/sending Auto Email Report with XLSX or CSV formats.

<img width="1552" alt="Screenshot 2021-11-01 at 12 47 36 PM" src="https://user-images.githubusercontent.com/36654812/139636250-cc0ff655-c943-4b05-8d09-53fae73060d5.png">

<!-- no-docs -->

gavindsouza committed:
feat(minor): Implement __getitem__ in Base Document

This addition allows for syntactic sugars like writing doc["name"] to
access document name. This addition is solely for the sake of code
simplicity. Since dictionaries in Python follow this standard, we can
write code that works for both Vanilla Dicts, Frappe Dicts & all
Document objects

gavindsouza commented:
@Mergifio backport develop version-13-pre-release

gavindsouza commented:
@Mergifyio backport develop version-13-pre-release

ankush commented on frappe/model/base_document.py:
```diff
@@ -85,6 +85,9 @@ def __init__(self, d):
 			self.__setup__()
 
+	def __getitem__(self, key):
+		return self.get(key)
```
This will return `None` for non-existing keys, and propagate it till it fails. Earlier it would fail immediately with `AttributeError`.

This usually makes debugging harder 🥲

gavindsouza replied:
Earlier this would fail with `TypeError: '{DocType}' object is not subscriptable`. Am I missing something here? 

---

Should this be 

```suggestion
		return self.get(key) or frappe.throw(msg=key, exc=KeyError)
```

instead?

ankush replied:
I am not talking about subscripting. Accessing keys in general.

`doctype.key` would throw AttributeError, but now doctype["key"] will just return `None`. This is not common behaviour as `Dict` throws key error. 


The suggested change is not ok either, it will fail when values are actually `None`, failure should only be there when key doesn't exist. Maybe`getattr` is better (but that's not identical to `Document.get` 🥲 ?

gavindsouza replied:
Okay, now I'm confused if we should raise KeyError or AttributeError if the attribute doesn't exist. I'm choosing `AttributeError` because although it feels like a dict, looks like a dict (sometimes), is a Document.

```suggestion
		return self.get(key) if hasattr(self, key) else frappe.throw(msg=key, exc=AttributeError)
```

ankush replied:
Isn't this `getattr(self, key)` with extra steps 😄

gavindsouza replied:
CRAY

```suggestion
		return getattr(self, key)
```

gavindsouza committed:
fix: Use getattr over get in getitem

For the sake of consistency and some level of sanity in the universe.


Co-authored-by: Ankush Menat <ankush@frappe.io>

gavindsouza committed:
fix: Raise KeyError instead of AttributeError

why: shit breaks down. Specifically stuff in Jinja templates, attributes
that are conditional. Either this is acceptable or subscripting objects
cannot and probably should not be possible along with our pre-condiitons

gavindsouza commented on frappe/model/base_document.py:
```diff
@@ -85,6 +85,9 @@ def __init__(self, d):
 			self.__setup__()
 
+	def __getitem__(self, key):
+		return self.get(key, frappe.throw(msg=key, exc=KeyError))
```
Looks better ig?

```suggestion
		return self.get(key) if hasattr(self, key) else frappe.throw(msg=key, exc=KeyError)
```

gavindsouza committed:
chore: Use hasattr instead of get default

Couldn't decide between a style commit or a chore
